modules/library/xml/src/main/java/org/geotools/xml/styling/SLDParser.java

```diff
@@ -201,80 +199,95 @@
 
     private boolean disposeInputSource;
 
-    private ExpressionDOMParser expressionDOMParser = new ExpressionDOMParser(FF);
+    private final ExpressionDOMParser expressionDOMParser;
 
     /**
-     * Create a Stylereader - use if you already have a dom to parse.
+     * Create a SLDParser.
      *
-     * @param factory The StyleFactory to use to build the style
+     * @param styleFactory The StyleFactory to use to build the style
+     * @param filterFactory The FilterFactory to create Filters.
+     * @throws java.lang.NullPointerException if any of the arguments are null.
      */
-    public SLDParser(StyleFactory factory) {
-        this(factory, CommonFactoryFinder.getFilterFactory(GeoTools.getDefaultHints()));
-    }
-
-    public SLDParser(StyleFactory factory, FilterFactory filterFactory) {
-        this.factory = factory;
+    public SLDParser(StyleFactory styleFactory, FilterFactory filterFactory) {
+        Objects.requireNonNull(styleFactory);
+        Objects.requireNonNull(filterFactory);
+        this.factory = styleFactory;
         this.ff = filterFactory;
+        this.expressionDOMParser = new ExpressionDOMParser(CommonFactoryFinder.getFilterFactory2());
```

Comment:
Shouldn't it use the filter factory provided as an argument, at least if that one is a FilterFactory2 (which should be most of the time, the only actual implementation in GeoTools implements both FF2 and thus FF).

Reply:
Actually with merge of #3516 it could be just 
`this.expressionDOMParser = new ExpressionDOMParser()`

However ExpressionDOMParser second constructor uses FilterFactory2 interface.

IMHO using filterFactory from SLDParser constructor arg assumes internal knowlege about the implementiontion of FilterFactoryImpl. But if someone else provides a FilterFactory via META-INF/services which doesn't implement both, FF and FF2 this class could not be used anymore. With other words a cast from FF to FF2 is from interface perspective not a good choise because you know details about implementation. This would not be a good example for the whole codebase because this kind of patterns ist copied and pasted and makes it hard to understand the code and to maintain as well. Looking forward for proposal @jodygarnett set up few years ago

Reply:
That's why I said "if". the code can be modified to test _if_ the provided factory is a FF2, and use it, otherwise create a different one. In the last 16 years spent working every day on GT/GS I have yet to see another implementation of FilterFactory, but yes, it may happen, and might even have aleady happened :-D